File: Java/Reorder-List.java
/**
 * Definition for ListNode.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int val) {
 *         this.val = val;
 *         this.next = null;
 *     }
 * }
 */ 
public class Solution {
    /**
     * @param head: The head of linked list.
     * @return: void
     */
    public ListNode FindMid(ListNode head) {
    	if (head == null) return null;
    	ListNode slow = head;
    	ListNode fast = head.next;
    	while (fast != null && fast.next != null) {
    		fast = fast.next.next;
    		slow = slow.next;
    	}
    	return slow;
    }
    public ListNode reverse(ListNode head) {
        // write your code here
        ListNode cur = null;
        ListNode cnt = head;
        while (cnt != null) {
        	ListNode temp = cnt.next;
        	cnt.next = cur;
        	cur = cnt;
        	cnt = temp;
        }
        return cur;
    }
    public void merge(ListNode head, ListNode start) {
    	ListNode cur = new ListNode(-1);
    	int index = 0;
    	while (head != null && start != null) {
    		if (index % 2 == 0) {
    			cur.next = head;
    			head = head.next;
    		} else {
    			cur.next = start;
    			start = start.next;
    		}
    		cur = cur.next;
    		index ++;
    	}
    	if (head != null) {
    		cur.next = head;
    	} else {
    		cur.next = start;
    	}
    }
    public void reorderList(ListNode head) {  
        // write your code here
        if (head == null || head.next == null) return;
    	ListNode mid = FindMid(head);
    	ListNode start = mid.next;
    	mid.next = null;
    	start = reverse(start);
    	merge(head, start);
    }
}

File: Cpp/Reorder-List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    void reorderList(ListNode *head) {
        // IMPORTANT: Please reset any member data you declared, as
        // the same Solution instance will be reused for each test case.
        if(head == NULL || head->next == NULL)return;
        ListNode *fastp = head, *lowp = head, *tail = NULL;
        while(fastp != NULL && fastp->next != NULL)
        {//利用快慢指针找到链表的中点
            tail = lowp;
            fastp = fastp->next->next;
            lowp = lowp->next;
        }
        tail->next = NULL; //此时tail 指向前半段的结尾
        reverseList(lowp);//翻转链表后半段
        fastp = head;
        tail = NULL;
        while(fastp != NULL)
        {
            ListNode *tmp = lowp->next;
            lowp->next = fastp->next;
            fastp->next = lowp;
            tail = lowp;
            fastp = lowp->next;
            lowp = tmp;
        }
        if(lowp != NULL)
            tail->next = lowp;

    }
    void reverseList(ListNode* &head)
    {//翻转链表
        if(head == NULL || head->next == NULL)return;
        ListNode *pre = head, *p = pre->next;
        while(p != NULL)
        {
            ListNode *tmp = p->next;
            p->next = pre;
            pre = p;
            p = tmp;
        }
        head->next = NULL;
        head = pre;
    }
};
